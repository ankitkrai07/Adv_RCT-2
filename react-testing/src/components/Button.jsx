import React from "react";
import PropTypes from "prop-types";

export const Button = ({ children, size, color }) => {
  return <button className={`${size} ${color}`}>{children}</button>;
};

Button.propTypes = {
  color: PropTypes.string.isRequired,
  size: PropTypes.string.isRequired,
};
