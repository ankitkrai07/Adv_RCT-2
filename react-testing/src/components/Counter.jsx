import React, { useState } from "react";
import { Button } from "./Button";

export const Counter = () => {
  const [count, setCount] = useState(0);

  const handleAdd = () => {
    setCount((prev) => prev + 1);
  };

  const handleReduce = () => {
    setCount((prev) => prev - 1);
  };

  return (
    <div>
      <h1 data-testid="counter">Counter:{count}</h1>
      {/* <button onClick={handleAdd}>ADD</button> */}
      <button onClick={handleReduce}>Reduce</button>
      <Button func={handleAdd} color="gray" size="small">
        Add
      </Button>
    </div>
  );
};
