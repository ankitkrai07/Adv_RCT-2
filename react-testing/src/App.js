import "./App.css";
import { Button } from "./components/Button";
import { Counter } from "./components/Counter";

function App() {
  return (
    <div className="App">
      <h1>React testing</h1>
      <Button size="medium" color="blue">
        Click Me
      </Button>
      <Counter />
    </div>
  );
}

export default App;
