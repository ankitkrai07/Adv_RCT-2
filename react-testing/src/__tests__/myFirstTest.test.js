import { fireEvent, render, screen } from "@testing-library/react";
import { Button } from "../components/Button";
import App from "../App";
import { Counter } from "../components/Counter";
import renderer from "react-test-renderer";

describe("Testing Button Component", () => {
  // Check button is present on dom or not
  it("Check button is present on dom or not", () => {
    render(<Button>Testing Button</Button>);
    screen.debug();
    // const button = screen.getByText("Test Button");
    // expect(button[0]).toBeInTheDocument();
  });

  it("Button is rendered on App with text Click Me", () => {
    render(<App />);
    const button = screen.getAllByTestId("test-button");
    expect(button[0]).toBeInTheDocument();
    expect(button[0]).toHaveTextContent("Click Me");
  });
  it("OnClicking the button the function should execute", () => {
    const mockFunc = jest.fn();
    render(<Button func={mockFunc}>Click</Button>);
    const button = screen.getByTestId("test-button");
    fireEvent.click(button);
    fireEvent.click(button);
    // expect(mockFunc).toBeCalled();
    // expect(mockFunc).toBeCalledTimes(2); ---> Problem here
  });

  it("Should throw error for color and size", () => {
    jest.spyOn(console, "error");
    render(
      <Button color={1} size={false}>
        Click
      </Button>
    );
    expect(console.error).toBeCalledTimes(2);
  });

  // Snapshot testing

  //   test("Link renders correctly", () => {
  //     const tree = renderer
  //       .create(<Link page="http://www.facebook.com">Facebook</Link>)
  //       .toJSON();
  //     expect(tree).toMatchSnapshot();
  //   });

  it("Link renders correctly", () => {
    const tree = renderer
      .create(
        <Button color="blue" size="large">
          Click
        </Button>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Checking counter functionalites", () => {
  it("Counter should be present on DOM with initial value", () => {
    render(<Counter />);
    const counter = screen.getByTestId("counter");
    expect(counter).toBeInTheDocument();
    expect(counter).toHaveTextContent(0);
  });

  it("On clicking add button counter should increase", () => {
    render(<Counter />);
    const counter = screen.getByTestId("counter");
    expect(counter).toHaveTextContent(0);
    const button = screen.getByTestId("test-button");
    fireEvent.click(button);
    expect(counter).toHaveTextContent(1);
    fireEvent.click(button);
    expect(counter).toHaveTextContent(2);
  });
});
