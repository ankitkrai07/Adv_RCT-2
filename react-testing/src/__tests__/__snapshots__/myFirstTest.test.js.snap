// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Button Component Link renders correctly 1`] = `
<button
  className="large blue"
  data-testid="test-button"
>
  Click
</button>
`;
